### PR TITLE
Broadcast arithmetic with `StepRange{ZonedDateTime}` returns a `StepRange`

### PR DESCRIPTION
**Pull request by omus**
In re-evaluating the performance of [this old benchmark](https://github.com/JuliaTime/TimeZones.jl/issues/182#issuecomment-459380018) I noticed the performance has actually gotten worse with the recent changes even though there are less allocations:

```julia
julia> using Dates, TimeZones, BenchmarkTools

julia> @btime (DateTime(2019):Day(1):DateTime(2020)) .+ Minute(1);
  309.455 ns (1 allocation: 3.00 KiB)

julia> @btime (ZonedDateTime(2019,tz"UTC"):Day(1):ZonedDateTime(2020,tz"UTC")) .+ Minute(1);
  7.023 μs (1 allocation: 14.44 KiB)
```

Upon investigating I was surprized that a `Vector` was returned instead of a `StepRange`:
```julia
julia> @btime (ZonedDateTime(2019,tz"UTC"):Day(1):ZonedDateTime(2020,tz"UTC")) .+ Minute(1)
  7.061 μs (1 allocation: 14.44 KiB)
366-element Array{ZonedDateTime,1}:
 2019-01-01T00:01:00+00:00
 2019-01-02T00:01:00+00:00
 2019-01-03T00:01:00+00:00
 ⋮
 2019-12-30T00:01:00+00:00
 2019-12-31T00:01:00+00:00
 2020-01-01T00:01:00+00:00
 ```

This PR changes arithmetic for `StepRange{ZonedDateTime}` to return a `StepRange` instead of a `Vector`:
```julia
julia> @btime (ZonedDateTime(2019,tz"UTC"):Day(1):ZonedDateTime(2020,tz"UTC")) .+ Minute(1)
  130.477 ns (0 allocations: 0 bytes)
ZonedDateTime(2019, 1, 1, 0, 1, tz"UTC"):Day(1):ZonedDateTime(2020, 1, 1, 0, 1, tz"UTC")
```

I'm also attempting to get this changed for `TimeType`s in general: https://github.com/JuliaLang/julia/pull/37759

### CODE REVIEW

**Comment by omus** [comment]
Using a `StepRange` instead of a `Vector` for broadcasted range arithmetic has significant performance benefits. We're only seeing an improvement for `TimePeriod` as `DatePeriod`s were already returning `StepRange`s.

# Benchmark Report for */Users/omus/.julia/dev/TimeZones*

## Job Properties
* Time of benchmarks:
    - Target: 28 Sep 2020 - 14:15
    - Baseline: 28 Sep 2020 - 14:16
* Package commits:
    - Target: cabc91
    - Baseline: 3800fc
* Julia commits:
    - Target: 539f3c
    - Baseline: 539f3c
* Julia command flags:
    - Target: None
    - Baseline: None
* Environment variables:
    - Target: None
    - Baseline: None

## Results
A ratio greater than `1.0` denotes a possible regression (marked with :x:), while a ratio less
than `1.0` denotes a possible improvement (marked with :white_check_mark:). Only significant results - results
that indicate possible regressions or improvements - are shown below (thus, an empty table means that all
benchmark results remained invariant between builds).

| ID                                                           | time ratio                   | memory ratio                 |
|--------------------------------------------------------------|------------------------------|------------------------------|
| `["ZonedDateTime", "fill"]`                                  | 0.86 (5%) :white_check_mark: |                   1.00 (1%)  |
| `["ZonedDateTime", "range", "FixedTimeZone/DatePeriod"]`     | 0.88 (5%) :white_check_mark: |                   1.00 (1%)  |
| `["ZonedDateTime", "range", "FixedTimeZone/TimePeriod"]`     | 0.88 (5%) :white_check_mark: |                   1.00 (1%)  |
| `["arithmetic", "broadcast", "FixedTimeZone/TimePeriod"]`    | 0.05 (5%) :white_check_mark: | 0.00 (1%) :white_check_mark: |
| `["arithmetic", "broadcast", "VariableTimeZone/TimePeriod"]` | 0.04 (5%) :white_check_mark: | 0.02 (1%) :white_check_mark: |

## Benchmark Group List
Here's a list of all the benchmark groups executed by this job:

- `["ZonedDateTime"]`
- `["ZonedDateTime", "local"]`
- `["ZonedDateTime", "range"]`
- `["arithmetic"]`
- `["arithmetic", "broadcast"]`
- `["interpret", "local"]`
- `["interpret"]`
- `["parse"]`
- `["transition_range", "local"]`
- `["transition_range"]`
- `["tryparsenext_fixedtz"]`
- `["tryparsenext_tz"]`
- `["tz_data"]`

## Julia versioninfo

### Target
```
Julia Version 1.5.2
Commit 539f3ce943 (2020-09-23 23:17 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin18.7.0)
  uname: Darwin 18.7.0 Darwin Kernel Version 18.7.0: Thu Jun 18 20:50:10 PDT 2020; root:xnu-4903.278.43~1/RELEASE_X86_64 x86_64 i386
  CPU: Intel(R) Core(TM) i7-7567U CPU @ 3.50GHz:
              speed         user         nice          sys         idle          irq
       #1  3500 MHz     636474 s          0 s     524032 s    4048186 s          0 s
       #2  3500 MHz     345356 s          0 s     190882 s    4672449 s          0 s
       #3  3500 MHz     652379 s          0 s     368730 s    4187577 s          0 s
       #4  3500 MHz     310069 s          0 s     159050 s    4739568 s          0 s

  Memory: 16.0 GB (112.625 MB free)
  Uptime: 521191.0 sec
  Load Avg:  3.85009765625  4.83203125  4.232421875
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-9.0.1 (ORCJIT, skylake)
```

### Baseline
```
Julia Version 1.5.2
Commit 539f3ce943 (2020-09-23 23:17 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin18.7.0)
  uname: Darwin 18.7.0 Darwin Kernel Version 18.7.0: Thu Jun 18 20:50:10 PDT 2020; root:xnu-4903.278.43~1/RELEASE_X86_64 x86_64 i386
  CPU: Intel(R) Core(TM) i7-7567U CPU @ 3.50GHz:
              speed         user         nice          sys         idle          irq
       #1  3500 MHz     636769 s          0 s     524155 s    4048444 s          0 s
       #2  3500 MHz     345570 s          0 s     190933 s    4672860 s          0 s
       #3  3500 MHz     652718 s          0 s     368838 s    4187806 s          0 s
       #4  3500 MHz     310263 s          0 s     159098 s    4740002 s          0 s

  Memory: 16.0 GB (193.26171875 MB free)
  Uptime: 521259.0 sec
  Load Avg:  3.392578125  4.5146484375  4.15478515625
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-9.0.1 (ORCJIT, skylake)
```

File: src/arithmetic.jl [hunk 1]
```diff
@@ -41,4 +41,10 @@ function broadcasted(::typeof(+), r::StepRange{ZonedDateTime}, p::DatePeriod)
     return StepRange(start, step, stop)
 end
 
-broadcasted(::typeof(-), r::StepRange{ZonedDateTime}, p::DatePeriod) = broadcast(+, r, -p)
+function broadcasted(::typeof(+), r::StepRange{ZonedDateTime}, p::TimePeriod)
+    StepRange(r.start + p, r.step, r.stop + p)
+end
+
+broadcasted(::typeof(+), p::Period, r::StepRange{ZonedDateTime}) = broadcasted(+, r, p)
+broadcasted(::typeof(-), r::StepRange{ZonedDateTime}, p::Period) = broadcasted(+, r, -p)
```
Comment (nicoleepp) [inline src/arithmetic.jl:49]: Curious about why the change from `DatePeriod` to `Period`

File: src/arithmetic.jl [hunk 2]
```diff
@@ -41,4 +41,10 @@ function broadcasted(::typeof(+), r::StepRange{ZonedDateTime}, p::DatePeriod)
     return StepRange(start, step, stop)
 end
 
-broadcasted(::typeof(-), r::StepRange{ZonedDateTime}, p::DatePeriod) = broadcast(+, r, -p)
+function broadcasted(::typeof(+), r::StepRange{ZonedDateTime}, p::TimePeriod)
+    StepRange(r.start + p, r.step, r.stop + p)
```
Comment (nicoleepp) [inline src/arithmetic.jl:45]:
```suggestion
    return StepRange(r.start + p, r.step, r.stop + p)
```

**Review by nicoleepp — APPROVED** [review]

File: src/arithmetic.jl (hunk 1 above)
Comment (omus) [reply src/arithmetic.jl:49]: Good question. Really I updated the code from:
```julia
broadcasted(::typeof(+), r::StepRange{ZonedDateTime}, p::DatePeriod) = ...
broadcasted(::typeof(-), r::StepRange{ZonedDateTime}, p::DatePeriod) = broadcasted(+, r, -p)
```
to
```julia
broadcasted(::typeof(+), r::StepRange{ZonedDateTime}, p::DatePeriod) = ...
broadcasted(::typeof(+), p::DatePeriod, r::StepRange{ZonedDateTime}) = broadcasted(+, r, p)
broadcasted(::typeof(-), r::StepRange{ZonedDateTime}, p::DatePeriod) = broadcasted(+, r, -p)
broadcasted(::typeof(-), p::DatePeriod, r::StepRange{ZonedDateTime}) = broadcasted(-, r, p)

broadcasted(::typeof(+), r::StepRange{ZonedDateTime}, p::TimePeriod) = ...
broadcasted(::typeof(+), p::TimePeriod, r::StepRange{ZonedDateTime}) = broadcasted(+, r, p)
broadcasted(::typeof(-), r::StepRange{ZonedDateTime}, p::TimePeriod) = broadcasted(+, r, -p)
broadcasted(::typeof(-), p::TimePeriod, r::StepRange{ZonedDateTime}) = broadcasted(-, r, p)
```
However, since `DatePeriod`/`TimePeriod` are subtypes of `Period` I can use a more generic signature to avoid duplicating methods. Doing that results in what's currently coded:
```julia
broadcasted(::typeof(+), r::StepRange{ZonedDateTime}, p::DatePeriod) = ...
broadcasted(::typeof(+), r::StepRange{ZonedDateTime}, p::TimePeriod) = ...

broadcasted(::typeof(+), p::Period, r::StepRange{ZonedDateTime}) = broadcasted(+, r, p)
broadcasted(::typeof(-), r::StepRange{ZonedDateTime}, p::Period) = broadcasted(+, r, -p)
broadcasted(::typeof(-), p::Period, r::StepRange{ZonedDateTime}) = broadcasted(-, r, p)
```
